### PR TITLE
Copter Mode enum class

### PR DESCRIPTION
**Pull request by khancyr**
This change Copter mode enum to enum class and move them to the mode class to limit their scope.

This is mostly a renaming, that doesn't bring changes. The enum class bring scope and type safety on compile time. The move into the class scope should reduce compile time.

Those change are covered by the compiler and autotest

### CODE REVIEW

**Comment by amilcarlucas** [comment]
Can the value of those enums be validated against Mavlink enum values using static asserts on a separated PR?

**Comment by khancyr** [comment]
those internal mode and aren't related to mavlink

**Review by rmackay9 — CHANGES_REQUESTED** [review]
In general it looks good but a few things I noticed:

- S-curves is getting close so I'd prefer we don't make any changes to mode_auto.cpp until that goes in
- Having classes called "ModeGuided" and "GuidedMode" seems confusing and although it is at least partially an existing issue I think we should call these sub modes.
- If we change from camel case to all caps or small let's use underscores between words.  I.e PathFollow should become PATH_FOLLOW or path_follow.

Txs!

**Comment by peterbarker** [comment]
On Tue, 26 Jan 2021, Randy Mackay wrote:

>  *  Having classes called "ModeGuided" and "GuidedMode" seems confusing and although it is at least partially an existing issue I think we should call
>     these sub modes.

I thought the same thing....

ModeGuided::SubMode::TAKEOFF?

**Comment by khancyr** [comment]
rebased, cleanup, switch to SubMode syntax and separate words !
I have remove Guided mode changes to not collide with https://github.com/ArduPilot/ardupilot/pull/17162

File: ArduCopter/mode_rtl.cpp [hunk 1]
```diff
@@ -187,10 +187,10 @@ void ModeRTL::climb_return_run()
     _state_complete = wp_nav->reached_wp_destination();
 }
 
-// rtl_loiterathome_start - initialise return to home
+// rtl_LOITER_AT_HOME_start - initialise return to home
```
Comment (peterbarker) [inline ArduCopter/mode_rtl.cpp:190]: Well, the initial comment was wrong - I'm not sure if this is wronger or not.

File: ArduCopter/mode_rtl.cpp [hunk 2]
```diff
@@ -377,10 +377,10 @@ void ModeRTL::land_start()
 
 bool ModeRTL::is_landing() const
 {
-    return _state == RTL_Land;
+    return _state == SubMode::LAND;
 }
 
-// rtl_returnhome_run - return home
+// rtl_RETURN_HOME_run - return home
```
Comment (peterbarker) [inline ArduCopter/mode_rtl.cpp:383]: Even more wrongerer.  Not even the name of the function here.

**Review by peterbarker — CHANGES_REQUESTED** [review]
I really do like this change.

... but it does show the downside of doing a global replace without checking every substitution.

**Comment by khancyr** [comment]
Comments rename corrected !

File: ArduCopter/mode_rtl.cpp [hunk 3]
```diff
@@ -377,11 +377,11 @@ void ModeRTL::land_start()
 
 bool ModeRTL::is_landing() const
 {
-    return _state == RTL_Land;
+    return _state == SubMode::LAND;
 }
 
-// rtl_returnhome_run - return home
-//      called by rtl_run at 100hz or more
+// land_run - return home
```
Comment (peterbarker) [inline ArduCopter/mode_rtl.cpp:383]:
```suggestion
// land_run - run the landing controllers to put the aircraft on the ground
```

**Review by peterbarker — CHANGES_REQUESTED** [review]
Comment incorrect.  Should fix it while we're here.

**Review by peterbarker — APPROVED** [review]
LGTM

**Comment by peterbarker** [comment]
Labelled it DevCall in case Randy doesn't see it before then

**Review by rmackay9 — CHANGES_REQUESTED** [review]
I think it's confusing to have different enums with the same name in each class especially because they are public.  I think we should either make the SubMode enum private or change its name to make it different for each mode.  I.e. RTLSubMode or make it private.

**Comment by peterbarker** [comment]
On Thu, 29 Apr 2021, Randy Mackay wrote:

> I think it's confusing to have different enums with the same name in each class especially because they are public. I think we should either make the SubMode enum private or change its name to make it different for each mode. I.e. RTLSubMode or make it private.

*Everywhere* outside - from an enforced, C++, perspective would have
to reference the enumeration values as ModeRTL::RTLSubMode::LAND - nobody 
outside ModeRTL can use its submode without having ModeRTL:: on the front. 
That's the advantage of having it inside the ModeRTL class definition, 
after all.

public/private is actually irrelevant here.  If you're not a function in 
ModeRTL then you can ONLY reference its SubMode by include "ModeRTL" in 
the enumeration value name.

**Comment by rmackay9** [comment]
I think there's still an argument for making the submode private because there should be no need for an external user to know about it.

**Comment by khancyr** [comment]
I think that won't be an issue to put the enum in private, I will do it

**Comment by peterbarker** [comment]
On Fri, 30 Apr 2021, Pierre Kancir wrote:

> I think that won't be an issue to put the enum in private, I will do it

Pretty sure you can't as places outside the class are using it.  But they 
*must* prefix it with ModeRTL:: so there's no confusion.

**Comment by khancyr** [comment]
Hum, we cannot put them into private anyways. And as Peter said that is scope constraint to class access so that shouldn't be an issue

**Review by rmackay9 — APPROVED** [review]
OK, sorry for the delay, approved thanks!

**Pull request merged**